### ISSUE TITLE
Don't push parts of inequality expressions to either side of join in RelationPlanner#visitJoin 

### ISSUE THREAD

**Issue by sopel39** [prestosql/presto#5777]
Consider query:
```
SELECT lineitem.orderkey, orders.orderkey 
FROM lineitem JOIN orders 
ON lineitem.orderkey < orders.orderkey + 1
```
Currently, `RelationPlanner#visitJoin ` will pushdown computation of `orders.orderkey + 1` to build side. This means that both `orders.orderkey` and `orders.orderkey + 1` will be kept in build table. However, `orders.orderkey + 1` is only needed to evaluate filter condition and can be evaluated in place.

**Comment by sopel39** [comment]
cc @martint

**Comment by erichwang** [comment]
This is definitely something that can be improved. It looks like way because the join planning initially evolved from the original hash join operator which only supported simple equi-join clauses, and thus needed to project out the exact join key values before the joining process. We could probably (A) have a targeted optimizer rule that delays projections until after cross joins specifically (we don't want to do this generally speaking for joins since we can't tell if joins are going to expand or contract data sets), or (B) I was thinking we could just properly implement the nested loops join node to properly handle the general join case (expression evaluation, and thus also supporting the outer join types) that it always needed to do. Assuming that makes sense, we might want to explore Option B since that might be the simpler long term approach then one-off optimizer rules.

**Comment by sopel39** [comment]
Predicate pushdown handles that case, so maybe we just keep filter node above `JoinNode` and let it do it's work. There doesn't seem to be a need for the same logic to be duplicated in `RelationPlanner` and `PredicatePushdown`.

**Comment by findepi** [comment]
> We could probably (A) have a targeted optimizer rule that delays projections until after cross joins specifically (we don't want to do this generally speaking for joins since we can't tell if joins are going to expand or contract data sets)


Good point. However, CROSS JOIN may be an example where you actually want to project eagerly.
Otherwise you compute the `symbol + 1`  projection as many times as there are other side's rows.

**Comment by erichwang** [comment]
> Predicate pushdown handles that case, so maybe we just keep filter node above `JoinNode` and let PPD do it's work. There doesn't seem to be a need for the same logic to be duplicated in `RelationPlanner` and `PredicatePushdown`.

I guess it's not so much something that overlaps PredicatePushDown, but rather the RelationPlanner does a few inconsistent (although technically correct) things depending on the inner vs outer, equi vs non-equi joins, etc -- and this exists because those segments largely landed in the earlier days. A combination of those things results in what is reported here.

> Good point. However, CROSS JOIN may be an example where you actually want to project eagerly.
Otherwise you compute the symbol + 1 projection as many times as there are other side's rows.

Yup, it comes down to whether we think the NxM subexpression evaluation will be more expensive than the M subexpression evaluation + NxM projection of extra column. I guess it can go either way depending on the context. That being said, I *think* we've generally taken an implicit policy of not eagerly projecting out downstream subexpressions until they are needed by an operator -- someone please correct me if in accurate. In this specific case, it does feel a bit weird in general sense to have the projection done early when this isn't really done anywhere else in the engine.
